File: python/sets.py
if __name__ == "__main__":
    a = {2, 1, 3}
    b = {6, 9, 2, 6}  # Duplicates are ignored

    print(f"Set A: {a}")
    print(f"Set B: {b}")

    print(f"Length of Set A: {len(a)}")
    print(f"Length of Set B: {len(b)}")

    print("\n")

    a.add(4)
    b.add(4)

    print(f"Set A (after adding): {a}")
    print(f"Set B (after adding): {b}")

    print("\n")

    print(f"First element in Set A: {list(a)[0]}")
    print(f"First element in Set B: {list(b)[0]}")

    print(f"Third element in Set A: {list(a)[2]}")
    print(f"Third element in Set B: {list(b)[2]}")

    print("\n")

    a.discard(6)
    b.discard(6)

    print(f"Set A (after deleting): {a}")
    print(f"Set B (after deleting): {b}")

    # Set Union
    union_set = a | b  # or a.union(b)

    # Set Intersection
    intersection_set = a & b  # or a.intersection(b)

    print("Union:", union_set)
    print("Intersection:", intersection_set)

File: c++/sets.cpp
#include <iostream>
#include <set> // Sets are built in to most languages by default
#include <iterator>

int main() {
    // Sets automatically sort themselves
    // Cannot modify items in the set once they are added

    std::set<int> a = {2, 1, 3};
    std::set<int> b = {6, 9, 2};

    // O(N)
    std::cout << "Set A: ";
    for (int x : a) {
        std::cout << x << " ";
    }
    std::cout << "\n";

    std::cout << "Set B: ";
    for (int x : b) {
        std::cout << x << " ";
    }
    std::cout << "\n";

    // O(log(N)
    a.insert(4);
    b.insert(4);

    std::set<int>::iterator a1 = a.begin();
    std::set<int>::iterator b1 = b.begin();

    std::cout << "First element in set A: " << *a1 << "\n";
    std::cout << "First element in set B: " << *b1 << "\n";

    std::set<int>::iterator a3 = next(a1, 2);
    std::set<int>::iterator b3 = next(a1, 2);

    std::cout << "Third element in set A: " << *a3 << "\n";
    std::cout << "Third element in set B: " << *b3 << "\n";

    std::cout << "\n";

    // O(log(N)
    b.erase(6);

    std::cout << "Set A: ";
    for (int x : a) {
        std::cout << x << " ";
    }
    std::cout << "\n";

    std::cout << "Set B: ";
    for (int x : b) {
        std::cout << x << " ";
    }
    std::cout << "\n";

    std::set<int> union_set;
    std::set<int> intersection_set;

    // Set Union
    std::set_union(a.begin(), a.end(), b.begin(), b.end(), std::inserter(union_set, union_set.begin()));

    // Set Intersection
    std::set_intersection(a.begin(), a.end(), b.begin(), b.end(), std::inserter(intersection_set, intersection_set.begin()));

    std::cout << "Union: ";
    for (int x : union_set) { 
        std::cout << x << " ";
    }
    std::cout << "\n";

    std::cout << "Intersection: ";
    for (int x : intersection_set) {
        std::cout << x << " ";
    }
    std::cout << "\n";

    return 0;
}